inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,154 @@
+                if tasks_data[courseid][user_task["taskid"]]["succeeded"]:
+                    del tasks_data[courseid][user_task["taskid"]]
+            # Remove outdated tasks and LTi courses or courses with no unfinished available tasks with deadline
+            for outdated_task in outdated_tasks[courseid]:
+                if outdated_task in tasks_data[courseid]:
+                    del tasks_data[courseid][outdated_task]
+            if (not any(task in tasks for task in tasks_data[courseid])) or course.is_lti():
+                del open_courses[course.get_id()]
+        my_render=RenderOrdered()
+        time_planner = ["7", "14", "30", "unlimited"]
+
+        # Sort the courses based on the most urgent task for each course
+        open_courses = OrderedDict( sorted(iter(open_courses.items()), key=lambda x: (sort_by_deadline(x[1], tasks_data[x[0]].keys())[0]).get_accessible_time().get_soft_end_date() ))
+
+        return self.template_helper.render("coming_tasks.html",
+                                           template_folder=PATH_TO_PLUGIN + "/templates/",
+                                           open_courses=open_courses,
+                                           tasks_data=tasks_data,
+                                           my_render=my_render,
+                                           time_planner=time_planner,
+                                           submissions=except_free_last_submissions)
+
+
+
+class RenderOrdered:
+    """ Class representing a render which is sent to the html file where it is recursively called with jinja """
+
+    def __init__(self):
+        pass
+
+    def order(self, course, task_list):
+        """ This method is encapsuled in the render object to allow calling easily from jinja """
+        return sort_by_deadline(course, task_list)
+
+    def render_upcoming_list(self, template_helper, course, tasks_data, render):
+        """ Used to render the task list filtered on deadlines
+        course is a course object
+        tasks_data is a dictionnary of data about tasks (including grade)
+        tasks_list is a list of urgent tasksid
+        the Render_Ordered object is sent to allow calling its order method """
+        return template_helper.render("upcoming.html", template_folder=PATH_TO_PLUGIN + '/templates/', course=course,
+                                      tasks_data=tasks_data, render=render)
+
+def sort_by_deadline(course, user_urgent_task_list):
+    """ Given a course (object) and a list of user urgent tasksid,
+    returns the list of urgent tasks (objects) for that course ordered based on deadline """
+    course_tasks = course.get_tasks()
+    course_user_urgent_task_list = list(set(course_tasks).intersection(user_urgent_task_list))
+    ordered_tasks = sorted(course_user_urgent_task_list, key=lambda x: course.get_task(x).get_accessible_time().get_soft_end_date())
+    ordered_tasks = map(course.get_task, ordered_tasks)
+    return list(ordered_tasks)
```

Comment:
You can probably replace the two last lines by:

    return [course_tasks[taskid] for taskid in ordered_tasks]

Reply:
Indeed, works fine.